File: python/basic_functions.py
"""
Basic Functions - GitHub Copilot Training Examples

This file demonstrates how to write clear, descriptive function names
and comments that help GitHub Copilot generate better code suggestions.
"""

import math
from typing import List, Dict, Optional


def calculate_circle_area(radius: float) -> float:
    """
    Calculate the area of a circle given its radius.
    
    Args:
        radius (float): The radius of the circle
        
    Returns:
        float: The area of the circle
    """
    return math.pi * radius ** 2


def convert_celsius_to_fahrenheit(celsius: float) -> float:
    """
    Convert temperature from Celsius to Fahrenheit.
    
    Args:
        celsius (float): Temperature in Celsius
        
    Returns:
        float: Temperature in Fahrenheit
    """
    return (celsius * 9/5) + 32


def find_largest_number_in_list(numbers: List[float]) -> Optional[float]:
    """
    Find the largest number in a list of numbers.
    
    Args:
        numbers (List[float]): List of numbers to search
        
    Returns:
        Optional[float]: The largest number, or None if list is empty
    """
    if not numbers:
        return None
    return max(numbers)


def count_word_frequency(text: str) -> Dict[str, int]:
    """
    Count the frequency of each word in a given text.
    
    Args:
        text (str): The input text to analyze
        
    Returns:
        Dict[str, int]: Dictionary with words as keys and frequencies as values
    """
    words = text.lower().split()
    frequency = {}
    for word in words:
        # Remove punctuation
        word = word.strip('.,!?;:"')
        frequency[word] = frequency.get(word, 0) + 1
    return frequency


def is_palindrome(text: str) -> bool:
    """
    Check if a string is a palindrome (reads the same forwards and backwards).
    
    Args:
        text (str): The text to check
        
    Returns:
        bool: True if the text is a palindrome, False otherwise
    """
    # Remove spaces and convert to lowercase
    cleaned_text = text.replace(' ', '').lower()
    return cleaned_text == cleaned_text[::-1]


def calculate_factorial(n: int) -> int:
    """
    Calculate the factorial of a non-negative integer.
    
    Args:
        n (int): The number to calculate factorial for
        
    Returns:
        int: The factorial of n
        
    Raises:
        ValueError: If n is negative
    """
    if n < 0:
        raise ValueError("Factorial is not defined for negative numbers")
    if n == 0 or n == 1:
        return 1
    return n * calculate_factorial(n - 1)


def generate_fibonacci_sequence(count: int) -> List[int]:
    """
    Generate a Fibonacci sequence with the specified number of terms.
    
    Args:
        count (int): Number of Fibonacci numbers to generate
        
    Returns:
        List[int]: List containing the Fibonacci sequence
    """
    if count <= 0:
        return []
    elif count == 1:
        return [0]
    elif count == 2:
        return [0, 1]
    
    sequence = [0, 1]
    for i in range(2, count):
        next_num = sequence[i-1] + sequence[i-2]
        sequence.append(next_num)
    
    return sequence


def validate_email_address(email: str) -> bool:
    """
    Validate if an email address has a basic valid format.
    
    Args:
        email (str): The email address to validate
        
    Returns:
        bool: True if email format is valid, False otherwise
    """
    import re
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return re.match(pattern, email) is not None


# Example usage and testing
if __name__ == "__main__":
    # Test the functions
    print("Circle area (radius=5):", calculate_circle_area(5))
    print("25°C in Fahrenheit:", convert_celsius_to_fahrenheit(25))
    print("Largest in [3, 1, 4, 1, 5]:", find_largest_number_in_list([3, 1, 4, 1, 5]))
    print("Word frequency in 'hello world hello':", count_word_frequency("hello world hello"))
    print("Is 'racecar' a palindrome?", is_palindrome("racecar"))
    print("Factorial of 5:", calculate_factorial(5))
    print("First 8 Fibonacci numbers:", generate_fibonacci_sequence(8))
    print("Is 'test@example.com' valid email?", validate_email_address("test@example.com"))

File: javascript/basic_functions.js
/**
 * Basic Functions - GitHub Copilot Training Examples
 * 
 * This file demonstrates how to write clear, descriptive function names
 * and comments that help GitHub Copilot generate better code suggestions.
 */

/**
 * Calculate the area of a circle given its radius
 * @param {number} radius - The radius of the circle
 * @returns {number} The area of the circle
 */
function calculateCircleArea(radius) {
    return Math.PI * radius ** 2;
}

/**
 * Convert temperature from Celsius to Fahrenheit
 * @param {number} celsius - Temperature in Celsius
 * @returns {number} Temperature in Fahrenheit
 */
function convertCelsiusToFahrenheit(celsius) {
    return (celsius * 9/5) + 32;
}

/**
 * Find the largest number in an array of numbers
 * @param {number[]} numbers - Array of numbers to search
 * @returns {number|null} The largest number, or null if array is empty
 */
function findLargestNumberInArray(numbers) {
    if (!numbers || numbers.length === 0) {
        return null;
    }
    return Math.max(...numbers);
}

/**
 * Count the frequency of each word in a given text
 * @param {string} text - The input text to analyze
 * @returns {Object} Object with words as keys and frequencies as values
 */
function countWordFrequency(text) {
    const words = text.toLowerCase()
        .replace(/[.,!?;:"]/g, '') // Remove punctuation
        .split(/\s+/)
        .filter(word => word.length > 0);
    
    const frequency = {};
    words.forEach(word => {
        frequency[word] = (frequency[word] || 0) + 1;
    });
    
    return frequency;
}

/**
 * Check if a string is a palindrome (reads the same forwards and backwards)
 * @param {string} text - The text to check
 * @returns {boolean} True if the text is a palindrome, false otherwise
 */
function isPalindrome(text) {
    const cleanedText = text.replace(/\s+/g, '').toLowerCase();
    return cleanedText === cleanedText.split('').reverse().join('');
}

/**
 * Calculate the factorial of a non-negative integer
 * @param {number} n - The number to calculate factorial for
 * @returns {number} The factorial of n
 * @throws {Error} If n is negative
 */
function calculateFactorial(n) {
    if (n < 0) {
        throw new Error('Factorial is not defined for negative numbers');
    }
    if (n === 0 || n === 1) {
        return 1;
    }
    return n * calculateFactorial(n - 1);
}

/**
 * Generate a Fibonacci sequence with the specified number of terms
 * @param {number} count - Number of Fibonacci numbers to generate
 * @returns {number[]} Array containing the Fibonacci sequence
 */
function generateFibonacciSequence(count) {
    if (count <= 0) return [];
    if (count === 1) return [0];
    if (count === 2) return [0, 1];
    
    const sequence = [0, 1];
    for (let i = 2; i < count; i++) {
        const nextNum = sequence[i-1] + sequence[i-2];
        sequence.push(nextNum);
    }
    
    return sequence;
}

/**
 * Validate if an email address has a basic valid format
 * @param {string} email - The email address to validate
 * @returns {boolean} True if email format is valid, false otherwise
 */
function validateEmailAddress(email) {
    const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    return emailRegex.test(email);
}

/**
 * Generate a random password with specified length and character types
 * @param {number} length - Length of the password
 * @param {Object} options - Options for character types to include
 * @param {boolean} options.includeUppercase - Include uppercase letters
 * @param {boolean} options.includeLowercase - Include lowercase letters  
 * @param {boolean} options.includeNumbers - Include numbers
 * @param {boolean} options.includeSymbols - Include symbols
 * @returns {string} Generated password
 */
function generateRandomPassword(length = 12, options = {}) {
    const {
        includeUppercase = true,
        includeLowercase = true,
        includeNumbers = true,
        includeSymbols = false
    } = options;
    
    let characters = '';
    if (includeUppercase) characters += 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    if (includeLowercase) characters += 'abcdefghijklmnopqrstuvwxyz';
    if (includeNumbers) characters += '0123456789';
    if (includeSymbols) characters += '!@#$%^&*()_+-=[]{}|;:,.<>?';
    
    if (characters === '') {
        throw new Error('At least one character type must be included');
    }
    
    let password = '';
    for (let i = 0; i < length; i++) {
        password += characters.charAt(Math.floor(Math.random() * characters.length));
    }
    
    return password;
}

/**
 * Format a number as currency
 * @param {number} amount - The amount to format
 * @param {string} currency - Currency code (default: 'USD')
 * @param {string} locale - Locale for formatting (default: 'en-US')
 * @returns {string} Formatted currency string
 */
function formatAsCurrency(amount, currency = 'USD', locale = 'en-US') {
    return new Intl.NumberFormat(locale, {
        style: 'currency',
        currency: currency
    }).format(amount);
}

// Example usage and testing
if (typeof window === 'undefined') { // Node.js environment
    console.log('Circle area (radius=5):', calculateCircleArea(5));
    console.log('25°C in Fahrenheit:', convertCelsiusToFahrenheit(25));
    console.log('Largest in [3, 1, 4, 1, 5]:', findLargestNumberInArray([3, 1, 4, 1, 5]));
    console.log('Word frequency in "hello world hello":', countWordFrequency("hello world hello"));
    console.log('Is "racecar" a palindrome?', isPalindrome("racecar"));
    console.log('Factorial of 5:', calculateFactorial(5));
    console.log('First 8 Fibonacci numbers:', generateFibonacciSequence(8));
    console.log('Is "test@example.com" valid email?', validateEmailAddress("test@example.com"));
    console.log('Random password:', generateRandomPassword(12, { includeSymbols: true }));
    console.log('Currency format:', formatAsCurrency(1234.56));
}

// Export functions for use in other modules (Node.js)
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        calculateCircleArea,
        convertCelsiusToFahrenheit,
        findLargestNumberInArray,
        countWordFrequency,
        isPalindrome,
        calculateFactorial,
        generateFibonacciSequence,
        validateEmailAddress,
        generateRandomPassword,
        formatAsCurrency
    };
}